crates/core/src/validity_predicate.rs

```diff
@@ -1,36 +1,32 @@
 //! Types that are used in validity predicates.
 
-/// Sentinel used in validity predicates to signal events that require special
-/// replay protection handling back to the protocol.
-#[derive(Debug, Default)]
-pub enum VpSentinel {
-    /// No action required
-    #[default]
-    None,
-    /// Exceeded gas limit
-    OutOfGas,
-    /// Found invalid transaction signature
-    InvalidSignature,
-}
+use thiserror::Error;
 
-impl VpSentinel {
-    /// Check if the Vp ran out of gas
-    pub fn is_out_of_gas(&self) -> bool {
-        matches!(self, Self::OutOfGas)
-    }
+use crate::borsh::{BorshDeserialize, BorshSerialize};
 
-    /// Check if the Vp found an invalid signature
-    pub fn is_invalid_signature(&self) -> bool {
-        matches!(self, Self::InvalidSignature)
-    }
+/// Helper trait for converting between result types.
+pub trait VpErrorExtResult<T> {
+    /// Convert to a [`Result`] with [`VpError`] errors.
+    fn into_vp_error(self) -> Result<T, VpError>;
+}
 
-    /// Set the sentinel for an out of gas error
-    pub fn set_out_of_gas(&mut self) {
-        *self = Self::OutOfGas
+impl<T, E> VpErrorExtResult<T> for Result<T, E>
+where
+    E: core::fmt::Display,
+{
+    // FIXME: we can't erase gas errors!
```

Comment:
Are we erasing them at the moment?

Reply:
we are, but I found out accidentally that it doesn't matter. when we exit out of a vp (either native or wasm), we add back the gas used by the vp to the tx's gas meter, which will cause that meter to also run out of gas. in the end, it doesn't really matter for us to distinguish between gas errors and other kinds of errors

Reply:
I think this works in case of a vp exceeding the transaction's gas limit but does it work even if the error is a gas (u64) overflow? I believe in this case the second addition would add the result of the overflow which probably would not trigger another overflow and could go unnoticed (even though we can expect an overflow error to be rare)